kyo-core/shared/src/main/scala/kyo/fibers.scala

```diff
@@ -171,24 +172,33 @@ object Fibers extends Joins[Fibers]:
                 }
 
     def race[T](l: Seq[T < Fibers])(using f: Flat[T < Fibers]): T < Fibers =
-        l.size match
-            case 0 => IOs.fail("Can't race an empty list.")
-            case 1 => l(0)
-            case _ =>
-                Fibers.get(raceFiber[T](l))
+        Fibers.get(raceFiber[T](l))
 
     def raceFiber[T](l: Seq[T < Fibers])(using f: Flat[T < Fibers]): Fiber[T] < IOs =
         l.size match
             case 0 => IOs.fail("Can't race an empty list.")
             case 1 => Fibers.run(l(0))
-            case _ =>
+            case size =>
                 Locals.save.map { st =>
                     IOs {
-                        val p = new IOPromise[T]
+                        class State extends AbstractFunction1[T < IOs, Unit]:
+                            val p       = new IOPromise[T]
+                            val pending = new AtomicInteger(size)
+                            def apply(v: T < IOs): Unit =
+                                val last = pending.decrementAndGet() == 0
+                                try discard(p.complete(IOs.run(v)))
+                                catch
+                                    case ex if (NonFatal(ex)) =>
```

Comment:
Do we need to branch on `NonFatal`? When would we not want to route errors through `IOs.fail`?

Reply:
Yes, fatal exceptions, which include interruptions, linkage errors, and OOMs are meant to immediately halt the execution of the current thread. A system can end up in a “zombie” state if that’s not the case.